master/internal/db/postgres.go

```diff
@@ -331,29 +334,36 @@ func (db *PgDB) queryRowsWithParser(
 			case reflect.Ptr:
 				sValue := reflect.New(vValue.Type().Elem().Elem())
 				if err = p(rows, sValue.Interface()); err != nil {
-					return err
+					return errors.Wrap(err, "queryRowsWithParser[ptr]")
 				}
 				vValue = reflect.Append(vValue, sValue)
 			case reflect.Struct:
 				sValue := reflect.New(vValue.Type().Elem())
 				if err = p(rows, sValue.Interface()); err != nil {
-					return err
+					return errors.Wrap(err, "queryRowsWithParser[struct]")
 				}
 				vValue = reflect.Append(vValue, sValue.Elem())
 			default:
 				return errors.Errorf("unexpected type: %s", k)
 			}
 		}
 		reflect.ValueOf(v).Elem().Set(vValue)
-		return nil
 	case reflect.Struct:
-		if rows.Next() {
-			return p(rows, v)
+		if !rows.Next() {
+			return ErrNotFound
+		}
+		if err = p(rows, v); err != nil {
+			return err
 		}
-		return ErrNotFound
 	default:
 		panic(fmt.Sprintf("unsupported query type: %s", kind))
 	}
+
+	if err := rows.Close(); err != nil {
```

Comment:
this is deferred on 326?

Reply:
right, but this one checks for errors.

The other one guarantees that the close happens.

@maxrussell told me the sql things have idempotent closers and that it is idiomatic to count on that.  That was a while ago, and I don't recall what he thought about this particular pattern.

Reply:
The idempotence of the method is noted in [the docs](https://pkg.go.dev/database/sql#Rows.Close). But I just tested with this:

```go
package main

import (
	"database/sql"
	"fmt"

	_ "github.com/lib/pq"
)

func main(){
	pool, err := sql.Open("postgres", "[redacted connection details]")
	if err != nil {
		panic(err)
	}

	rows, err := pool.Query("SELECT 1 UNION SELECT 2")
	if err != nil {
		panic(err)
	}

	for rows.Next() {
		var foo int
		err := rows.Scan(&foo)
		if err != nil {
			panic(err)
		}

		fmt.Println("scanned", foo)
	}
	if err := rows.Err(); err != nil {
		panic(err)
	} else {
		fmt.Println("No rows.Err()")
	}

	err = rows.Close()
	fmt.Println(err)
	
	err = rows.Close()
	fmt.Println(err)
}
```
and got the following output:
```
scanned 1
scanned 2
No rows.Err()
<nil>
<nil>
```

So the last two errors from closing resulted in the same `nil`.

That said, and apologies if you and I already covered this and I don't remember, @rb-determined-ai, but this could be a good use case for a named return and a deferred call to an anonymous function:

```go
func (db *PgDB) queryRowsWithParser(
	query string, p func(*sqlx.Rows, interface{}) error, v interface{}, args ...interface{},
) (returnedErr error) {
	rows, err := db.sql.Queryx(query, args...)
	if err != nil {
		return err
	}
	defer func(){
		closeErr = rows.Close()
		if returnedErr == nil && closeErr != nil {
			returnedErr = errors.Wrapf(closeErr, "rows.Close()")
		}
	}()
```

Personally, I find both to be equally readable and idiomatic. The existing version (with deferred `Close` to ensure cleanup and then later `Close` for return value) might be clearer with a comment to explain the reason for the double close, but again I don't have strong feelings either way.

Reply:
Well I need _somebody_ to care enough to either approve or ask it to be rewritten differently.  @stoksc or @maxrussell.

I put up a PR with a rather arbitrarily-chosen design choice.  I'm happy to fix it however you like.  The only thing I don't want is to leave room for a genuine bug, even a rare one.